### ISSUE TITLE
[AArch64][ComplexDeinterleaving] Miscompile: identifyReassocNodes drops a repeated addend, lowering acc + 2*x as acc + x

### ISSUE THREAD

**Issue by alexey-bataev** [llvm/llvm-project#205197]
ComplexDeinterleavingPass miscompiles a reassociated floating-point reduction. When an interleaved (complex) reduction adds each element twice, e.g. acc = acc + x; acc = acc + x (i.e. acc + 2*x), the pass rewrites it to add the element only once (acc + x), producing a wrong result.

In identifyReassocNodes, the Collect lambda flattens the add/sub/mul tree into an addend list:

```c++
SmallPtrSet<Value *, 8> Visited;
while (!Worklist.empty()) {
  auto [V, IsPositive] = Worklist.pop_back_val();
  if (!Visited.insert(V).second)
    continue;                                  // drops repeated addends
  ...
  if (I != Insn && I->hasNUsesOrMore(2)) {
    Addends.emplace_back(I, IsPositive);       // shared value added as a leaf
    continue;
  }
```

The deinterleave value x (%strided.vec) occurs twice in acc + (acc_prev + x)-style sums, i.e. it is a legitimate repeated addend. Visited deduplicates by value identity, so the second occurrence is skipped and the collected list is {x, acc} instead of {x, x, acc}. identifyAdditions then rebuilds acc + x instead of acc + 2*x. For reassoc FP, x + x is 2*x, not x, so dropping the occurrence is a miscompile. The RealAddends.size() == ImagAddends.size() guard does not catch it because both the real and imag chains lose one term symmetrically.

Suggested fix:

```c++
   auto Collect = [&Flags](Instruction *Insn, SmallVectorImpl<Product> &Muls,
                           AddendList &Addends) -> bool {
     SmallVector<PointerIntPair<Value *, 1, bool>> Worklist = {{Insn, true}};
-    SmallPtrSet<Value *, 8> Visited;
     while (!Worklist.empty()) {
       auto [V, IsPositive] = Worklist.pop_back_val();
-      if (!Visited.insert(V).second)
-        continue;
 
       Instruction *I = dyn_cast<Instruction>(V);
```

https://godbolt.org/z/75ravGxEd
